Scripts/Core/EditorVR.Menus.cs

```diff
@@ -197,119 +109,282 @@ internal void UpdateMenuVisibilities()
-				for (int i = 0; i < m_ActiveDeviceData.Count; i++)
+				var directSelection = evr.GetNestedModule<DirectSelection>();
+
+				foreach (var deviceData in m_ActiveDeviceData)
 				{
-					var deviceData = m_ActiveDeviceData[i];
 					var alternateMenu = deviceData.alternateMenu;
 					var mainMenu = deviceData.mainMenu;
 					var customMenu = deviceData.customMenu;
 					var menuHideFlags = deviceData.menuHideFlags;
 
-					// Move alternate menu to another device if it conflicts with main or custom menu
-					if (alternateMenu != null && (menuHideFlags[mainMenu] == 0 || (customMenu != null && menuHideFlags[customMenu] == 0)) && menuHideFlags[alternateMenu] == 0)
+					var mainMenuVisible = mainMenu != null && (menuHideFlags[mainMenu] & MenuHideFlags.Hidden) == 0;
+					var customMenuVisible = customMenu != null && (menuHideFlags[customMenu] & MenuHideFlags.Hidden) == 0 && (menuHideFlags[customMenu] & MenuHideFlags.OtherMenu) == 0;
+					var alternateMenuVisible = alternateMenu != null && (menuHideFlags[alternateMenu] & MenuHideFlags.Hidden) == 0;
+
+					if (alternateMenuVisible && (mainMenuVisible || customMenuVisible))
 					{
-						for (int j = 0; j < m_ActiveDeviceData.Count; j++)
+						foreach (var otherDeviceData in m_ActiveDeviceData)
 						{
-							var otherDeviceData = m_ActiveDeviceData[j];
 							if (otherDeviceData == deviceData)
 								continue;
 
-							var otherCustomMenu = otherDeviceData.customMenu;
-							var otherHideFlags = otherDeviceData.menuHideFlags;
-							otherHideFlags[otherDeviceData.alternateMenu] &= ~MenuHideFlags.Hidden;
-							otherHideFlags[otherDeviceData.mainMenu] |= MenuHideFlags.Hidden;
+							SetAlternateMenuVisibility(otherDeviceData.rayOrigin, true);
+							break;
+						}
+					}
 
-							if (otherCustomMenu != null)
-								otherHideFlags[otherCustomMenu] |= MenuHideFlags.Hidden;
+					if (customMenuVisible && (mainMenuVisible || alternateMenuVisible))
+						menuHideFlags[customMenu] |= MenuHideFlags.OtherMenu;
+
+					Array.Clear(m_WorkspaceOverlaps, 0, m_WorkspaceOverlaps.Length);
+					var hoveringWorkspace = false;
+					var menuTransform = mainMenu.menuContent.transform;
+					var menuBounds = mainMenu.localBounds;
+					var menuRotation = menuTransform.rotation;
+					var viewerScale = Viewer.GetViewerScale();
+					var center = menuTransform.position + menuRotation * menuBounds.center * viewerScale;
+					if (Physics.OverlapBoxNonAlloc(center, menuBounds.extents * viewerScale, m_WorkspaceOverlaps, menuRotation) > 0)
+					{
+						foreach (var overlap in m_WorkspaceOverlaps)
+						{
+							if (overlap)
+							{
+								m_WorkspaceComponents.Clear();
+								overlap.GetComponents(m_WorkspaceComponents);
+								if (m_WorkspaceComponents.Count > 0)
+									hoveringWorkspace = true;
+							}
 						}
+					}
 
-						menuHideFlags[alternateMenu] |= MenuHideFlags.Hidden;
+					var menus = menuHideFlags.Keys.ToList();
```

Comment:
Right. So this is a somewhat rare case where we're modifying each value-type value of a dictionary while iterating over it. We need to copy *something* in order to do this, because this doesn't work:

foreach (var kvp in dictionary)
    dictionary[kvp.Key] = someValue;

You get an InvalidOperationException.

My current solution is to keep a static list around and do this:

k_Menus.Clear();
k_Menus.AddRange(menuHideFlags.Keys);
foreach (var menu in k_Menus)
{
	menuHideFlags[menu] &= ~MenuHideFlags.Temporary;
}

This still allocates on KeyCollection.CopyTo, but it allocates less. I believe the only way to eliminate allocation entirely is to iterate twice, like so:

k_Menus.Clear();
foreach (var kvp in menuHideFlags)
{
	k_Menus.Add(kvp.Key);
}

foreach (var menu in k_Menus)
{
	menuHideFlags[menu] &= ~MenuHideFlags.Temporary;
}

This makes me think we should just keep a separate list of menus per device data, or maybe split the dictionary into two lists, and do a lookup only when needed? Curious if you have any alternate solutions in mind.